File: tests/protocol/tools/consume-built-origin.ts
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { createRemoteSkills } from "../../../packages/sdk-typescript/src/index.ts";
import type { ActivatedSessionSkill } from "../../../packages/sdk-typescript/src/session/index.ts";
import {
  decodeActivationExpectations,
  type ExpectedActivation,
} from "../helpers/activation-expectations.ts";

interface ConsumerOptions {
  cache: "off" | "on";
  cacheDirectory: string;
  expected: string;
  format: string;
  origin: string;
}

function requiredOption(options: ReadonlyMap<string, string>, name: string): string {
  const value = options.get(name);
  if (value === undefined) throw new Error(`missing --${name}`);
  return value;
}

function options(arguments_: readonly string[]): Readonly<ConsumerOptions> {
  const parsed = new Map<string, string>();
  for (let index = 0; index < arguments_.length; index += 2) {
    const name = arguments_[index];
    const value = arguments_[index + 1];
    if (!name?.startsWith("--") || value === undefined) {
      throw new Error("consumer arguments must be --name value pairs");
    }
    parsed.set(name.slice(2), value);
  }
  const cache = requiredOption(parsed, "cache");
  if (cache !== "off" && cache !== "on") {
    throw new Error("--cache must be off or on");
  }
  return Object.freeze({
    cache,
    cacheDirectory: requiredOption(parsed, "cache-dir"),
    expected: requiredOption(parsed, "expected"),
    format: requiredOption(parsed, "format"),
    origin: requiredOption(parsed, "origin"),
  });
}

function expectedActivation(value: ExpectedActivation) {
  const { outcome, requests, ...activation } = value;
  assert.equal(outcome, "activation_success");
  assert.equal(requests, 2);
  return activation;
}

async function normalizedActivation(skill: ActivatedSessionSkill) {
  return {
    origin_alias: skill.originAlias,
    name: skill.name,
    digest: skill.digest,
    instructions: skill.instructions,
    frontmatter: skill.frontmatter,
    files: await skill.list(),
  };
}

const selected = options(process.argv.slice(2));
const parsedExpected: unknown = JSON.parse(await readFile(resolve(selected.expected), "utf8"));
const expectedDocument = decodeActivationExpectations(parsedExpected);
const expected = expectedDocument.cases
  .filter(({ id }) => id.startsWith(`${selected.format}-`))
  .map(({ result }) => expectedActivation(result));
assert.equal(expected.length, 2, "the reviewed format must have two activation expectations");

for (let pass = 0; pass < 2; pass += 1) {
  const client = createRemoteSkills({
    origins: {
      "fixture-publisher": {
        url: selected.origin,
        allowLoopbackHttp: true,
        retries: 0,
      },
    },
    ...(selected.cache === "on"
      ? { cache: "disk", cacheOptions: { directory: resolve(selected.cacheDirectory) } }
      : { cache: "memory" }),
  });
  const session = await client.session("fixture-publisher");
  try {
    assert.equal((await session.catalog()).length, expected.length);
    const actual: Awaited<ReturnType<typeof normalizedActivation>>[] = [];
    for (const activation of expected) {
      actual.push(await normalizedActivation(await session.activate(activation.name)));
    }
    assert.deepEqual(actual, expected);
  } finally {
    await session.close();
  }
}

process.stdout.write(
  `${JSON.stringify({ language: "typescript", cache: selected.cache, passes: 2 })}\n`,
);

File: tests/protocol/tools/consume_built_origin.py
"""Consume one built publisher origin twice through the public Python SDK."""

from __future__ import annotations

import argparse
import asyncio
import json
from pathlib import Path

from remote_skills import Origin, RemoteSkills
from remote_skills.activation import verify_cached_archive
from remote_skills.cache import DiskCache, MemoryCache


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--origin", required=True)
    parser.add_argument("--format", required=True, choices=("tar-gzip", "zip"))
    parser.add_argument("--cache", required=True, choices=("off", "on"))
    parser.add_argument("--cache-dir", required=True, type=Path)
    parser.add_argument("--expected", required=True, type=Path)
    return parser.parse_args()


def expected_activation(value: dict[str, object]) -> dict[str, object]:
    if value.get("outcome") != "activation_success" or value.get("requests") != 2:
        raise AssertionError("reviewed activation expectation is malformed")
    return {
        key: item for key, item in value.items() if key not in {"outcome", "requests"}
    }


async def normalized_activation(skill: object) -> dict[str, object]:
    files = await skill.list()
    return {
        "origin_alias": skill.origin_alias,
        "name": skill.name,
        "digest": skill.digest,
        "instructions": skill.instructions,
        "frontmatter": dict(skill.frontmatter),
        "files": [
            {"path": item.path, "size": item.size, "media_type": item.media_type}
            for item in files
        ],
    }


async def main() -> None:
    arguments = parse_arguments()
    document = json.loads(arguments.expected.read_text(encoding="utf-8"))
    expected = [
        expected_activation(case["result"])
        for case in document["cases"]
        if case["id"].startswith(f"{arguments.format}-")
    ]
    if len(expected) != 2:
        raise AssertionError("the reviewed format must have two activation expectations")

    for _pass in range(2):
        cache = (
            DiskCache(arguments.cache_dir, archive_verifier=verify_cached_archive)
            if arguments.cache == "on"
            else MemoryCache(archive_verifier=verify_cached_archive)
        )
        client = RemoteSkills(
            origins={
                "fixture-publisher": Origin(
                    url=arguments.origin,
                    retries=0,
                    allow_loopback_http=True,
                )
            },
            cache=cache,
        )
        async with client.session("fixture-publisher") as session:
            if len(await session.catalog()) != len(expected):
                raise AssertionError("built catalog inventory differs from reviewed expectations")
            actual = [
                await normalized_activation(await session.activate(item["name"]))
                for item in expected
            ]
            if actual != expected:
                raise AssertionError("Python activation differs from reviewed expectations")

    print(json.dumps({"language": "python", "cache": arguments.cache, "passes": 2}))


if __name__ == "__main__":
    asyncio.run(main())
